app/controllers/sessions_controller.rb

```diff
@@ -0,0 +1,25 @@
+class SessionsController < ApplicationController
+  skip_before_action :verify_authenticity_token
+
+  def create
+    user = User.from_omniauth(request.env['omniauth.auth'])
+
+    session[:user_id] = user.id
+
+    redirect_to(
+      success_url,
+      notice: t('omniauth_callbacks.success')
+    )
+  end
+
+  def failure
+    flash[:error] = t('omniauth_callbacks.failure', reason: failure_message)
+    redirect_to root_url
+  end
+
+  private
+
+  def failure_message
+    env['omniauth.error.type'].to_s.humanize
```

Comment:
I'm fine with either the generic message or the one based on the omniauth error, which seems to be the convention based on what Devise does and other documentation I've seen.